### PR TITLE
MAINT: release branch PROPACK switch (default off)

### PR DESCRIPTION
**Pull request by tylerjereddy**
* on release branch `USE_PROPACK` env variable must now be set to a
non-zero value for SciPy to leverage `PROPACK` due
to ongoing challenges at:
https://github.com/conda-forge/scipy-feedstock/pull/198

* we could technically have it default-on for everything
but Windows, though if you'd prefer that then maybe help
me test the different scenarios if we add that extra logic
(different platform-specific defaults might be annoying though..)

* I've tested on MacOS as follows:

- `python runtests.py -t "scipy/sparse/linalg/" -- -n 6 -rsx`
  - `562 passed, 273 skipped, 8 xfailed in 5.70s`
- `USE_PROPACK=1 python runtests.py -t "scipy/sparse/linalg/" -- -n 6
  -rsx`
  - `715 passed, 159 skipped, 8 xfailed in 7.64s`

* we've given this some time, but we just don't have enough
bandwidth to have `PROPACK` default-on just yet; there
is one open main repo PR related this:
https://github.com/scipy/scipy/pull/15249

* adjust release notes accordingly

Maybe @mckib2 can take a look at the switch implementation details, and @rgommers for the general policy/approach and release notes description changes.

### CODE REVIEW

**Comment by tylerjereddy** [comment]
cc @h-vetinari

File: scipy/sparse/linalg/setup.py [hunk 1]
```diff
@@ -11,7 +12,8 @@ def configuration(parent_package='', top_path=None):
     config.add_data_dir('tests')
 
     # PROPACK
-    config.add_subpackage('_propack')
+    if os.environ.get("USE_PROPACK"):
+        config.add_subpackage('_propack')
```
Comment (rgommers) [inline scipy/sparse/linalg/setup.py:16]: This will break the opt-in for users. We can always build PROPACK, it should only be disabled at runtime. So please undo this change.

File: scipy/sparse/linalg/tests/test_propack.py [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import numpy as np
 from numpy.testing import assert_allclose
 from pytest import raises as assert_raises
+pytest.importorskip("scipy.sparse.linalg._svdp")
```
Comment (rgommers) [inline scipy/sparse/linalg/tests/test_propack.py:8]: This change should not be here either.

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 3]
```diff
@@ -6,7 +7,8 @@
 from scipy._lib._util import check_random_state
 from scipy.sparse.linalg._interface import LinearOperator, aslinearoperator
 from scipy.sparse.linalg._eigen.lobpcg import lobpcg  # type: ignore[no-redef]
-from scipy.sparse.linalg._svdp import _svdp
+if os.environ.get("USE_PROPACK"):
+    from scipy.sparse.linalg._svdp import _svdp
```
Comment (rgommers) [inline scipy/sparse/linalg/_eigen/_svds.py:11]: Can you add a check under `elif solver == 'propack':`, something like:
```python
HAS_PROPACK = False

....

if not HAS_PROPACK:
    raise ValueError("`solver='propack'` is opt-in due to potential issues on Windows, "
                      "it can be enabled by setting the `USE_PROPACK` environment "
                      "variable before importing scipy")

File: doc/release/1.8.0-notes.rst [hunk 4]
```diff
@@ -137,7 +139,9 @@ nonsingular non-Hermitian linear systems in `scipy.sparse.linalg.tfqmr`.
 
 The sparse SVD library PROPACK is now vendored with SciPy, and an interface is
 exposed via `scipy.sparse.svds` with ``solver='PROPACK'``. For some problems,
-this may be faster and/or more accurate than the default, ARPACK.
+this may be faster and/or more accurate than the default, ARPACK. PROPACK
+functionality is currently opt-in--you must specify ``USE_PROPACK=1`` at
+source build time to use it.
```
Comment (rgommers) [inline doc/release/1.8.0-notes.rst:144]: Add "due to potential issues on Windows that we aim to resolve in the next release".

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @tylerjereddy. Overall looks good to me, it needs a few tweaks.

**Comment by mckib2** [comment]
Thanks for getting this ready, unfortunately I'm not able to review this week. After a quick glance, I don't have anything to add on top of @rgommers review, the content and approach look good to be y

File: scipy/sparse/linalg/tests/test_propack.py (hunk 2 above)
Comment (tylerjereddy) [reply scipy/sparse/linalg/tests/test_propack.py]: Now there are no changes for this file, though I'm a bit confused--it seems to me like a runtime skip guard on the env variable for the module is still warranted here, similar to `pytestmark = pytest.mark.skipif(not os.environ.get(..`

Comment (rgommers) [reply scipy/sparse/linalg/tests/test_propack.py]: Probably not needed since each individual test already skips correctly and the import will always work, but it doesn't matter much either way.

File: scipy/sparse/linalg/_eigen/_svds.py (hunk 3 above)
Comment (rgommers) [reply scipy/sparse/linalg/_eigen/_svds.py:11]: This looks good to me.

**Review by rgommers — APPROVED** [review]
Okay looks great, let's get it in. Thanks @tylerjereddy!

**Comment by rgommers** [comment]
I guess this was the last piece of the puzzle for `1.8.0rc3`?

**Comment by tylerjereddy** [comment]
yeah, I guess the docs server stuff is going to be "testing in production"

**Comment by tylerjereddy** [comment]
ok, wheels are building, I'll let that go overnight

**Comment by tylerjereddy** [comment]
Just a quick note that the Windows wheel builds all failed so I'll need to investigate that later today.

I see Pythran stuff near the end of one of the full logs ( https://ci.appveyor.com/project/scipy/scipy-wheels/builds/42265533/job/caciav64t6l5fequ?fullLog=true ), so maybe there is still something I forgot to pin there, if I'm lucky.

File: scipy/sparse/linalg/tests/test_propack.py (hunk 2 above)
Comment (tylerjereddy) [reply scipy/sparse/linalg/tests/test_propack.py]: @rgommers @h-vetinari Are we sure about this? The 1.8.0rc3 OSX failures over at conda-forge in https://github.com/conda-forge/scipy-feedstock/pull/198 have tests suites that go like this:

```
...
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-single-array] PASSED [ 63%]
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-single-csr_matrix] PASSED [ 63%]
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-single-csc_matrix] PASSED [ 63%]
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-double-array] PASSED [ 63%]
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-double-csr_matrix] PASSED [ 63%]
sparse/linalg/tests/test_propack.py::test_svdp[LM-True-double-csc_matrix] PASSED [ 63%]
Fatal Python error: Segmentation fault

Current thread 0x000000010f49cdc0 (most recent call first):

```

Shouldn't those be skipping when the environment variable is not set? That's what worried me here--I didn't understand/see how these particular tests were autoskipping without some patch on our end.

**Comment by serge-sans-paille** [comment]
@tylerjereddy : the pythran error is due to cl.exe being used instead of clang-cl.exe. clang-cl is the default in pythran since 0.9.9 (and the log shows 0.10.0) but pythran is also sensible to CC and CXX environment variables, that may be an hint

**Comment by rgommers** [comment]
@serge-sans-paille we were clear on the hint, it has just not worked to convince the build system to ping up `clang.exe`. After an Azure image update (which dragged in the `vc142` toolchain) our monkeypatching house of cards seems to have collapsed.

Given that those problems completely go away when we switch to Meson, I gave up further debugging.

**Comment by serge-sans-paille** [comment]
ok, thanks @rgommers for the details, happy to wait and see rather than debugging that ;-)
